src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,329 @@
+/**
+ * @file genome.hpp
+ * @author Bang Liu
+ *
+ * Definition of the Genome class.
+ */
+#ifndef MLPACK_METHODS_NE_GENOME_HPP
+#define MLPACK_METHODS_NE_GENOME_HPP
+
+#include <cstddef>
+#include <cassert>
+#include <map>
+
+#include <mlpack/core.hpp>
+
+#include "link_gene.hpp"
+#include "neuron_gene.hpp"
+#include "utils.hpp"
+
+namespace mlpack {
+namespace ne {
+
+/**
+ * This class defines a genome.
+ # A genome is consist of a group of genes.
+ */
+class Genome {
+ public:
+  // Neurons.
+  std::vector<NeuronGene> aNeuronGenes;
+
+  // Links.
+  std::vector<LinkGene> aLinkGenes;
+
+  // Default constructor.
+  Genome() {}
+  
+  // Parametric constructor.
+  Genome(ssize_t id,
+  	     const std::vector<NeuronGene>& neuronGenes,
+         const std::vector<LinkGene>& linkGenes,
+         ssize_t numInput,
+         ssize_t numOutput,
+         double fitness,
+         double adjustedFitness):
+    aId(id),
+    aNeuronGenes(neuronGenes),
+    aLinkGenes(linkGenes),
+    aNumInput(numInput),
+    aNumOutput(numOutput),
+    aFitness(fitness),
+    aAdjustedFitness(adjustedFitness)
+  {}
+
+  // Copy constructor.
+  Genome(const Genome& genome) {
+    aId = genome.aId;
+    aNeuronGenes = genome.aNeuronGenes;
+    aLinkGenes = genome.aLinkGenes;
+    aNumInput = genome.aNumInput;
+    aNumOutput = genome.aNumOutput;
+    aFitness = genome.aFitness;
+    aAdjustedFitness = genome.aAdjustedFitness;
+  }
+
+  // Destructor.
+  ~Genome() {}
+
+  // Operator =.
+  Genome& operator =(const Genome& genome) {
+    if (this != &genome) {
+      aId = genome.aId;
+      aNeuronGenes = genome.aNeuronGenes;
+      aLinkGenes = genome.aLinkGenes;
+      aNumInput = genome.aNumInput;
+      aNumOutput = genome.aNumOutput;
+      aFitness = genome.aFitness;
+      aAdjustedFitness = genome.aAdjustedFitness;
+    }
+
+    return *this;
+  }
+
+  // Get genome id.
+  ssize_t Id() const { return aId; }
+
+  // Set genome id.
+  void Id(ssize_t id) { aId = id; }
+
+  // Get input length.
+  ssize_t NumInput() {
```

Comment:
Since the input number of inputs never changed and we don't add another Bias neuron, we could calculate NumInput once. It's a relative simple function, but we call it fairly often.